// react-stockcharts/lib/pure.js
"use strict";

Object.defineProperty(exports, "__esModule", {
	value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

var _get = function get(_x, _x2, _x3) { var _again = true; _function: while (_again) { var object = _x, property = _x2, receiver = _x3; desc = parent = getter = undefined; _again = false; if (object === null) object = Function.prototype; var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { _x = parent; _x2 = property; _x3 = receiver; _again = true; continue _function; } } else if ("value" in desc) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } } };

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var _react = require("react");

var _react2 = _interopRequireDefault(_react);

var _utilsShallowEqual = require("./utils/shallowEqual");

var _utilsShallowEqual2 = _interopRequireDefault(_utilsShallowEqual);

function getDisplayName(Series) {
	var name = Series.displayName || Series.name || "Series";
	return name;
}

function pure(PureSeries, contextShape) {
	var PureCanvasSeries = (function (_React$Component) {
		_inherits(PureCanvasSeries, _React$Component);

		function PureCanvasSeries() {
			_classCallCheck(this, PureCanvasSeries);

			_get(Object.getPrototypeOf(PureCanvasSeries.prototype), "constructor", this).apply(this, arguments);
		}

		_createClass(PureCanvasSeries, [{
			key: "shouldComponentUpdate",
			value: function shouldComponentUpdate(nextProps, nextState, nextContext) {
				return !(0, _utilsShallowEqual2["default"])(this.props, nextProps) || !(0, _utilsShallowEqual2["default"])(this.state, nextState) || !(0, _utilsShallowEqual2["default"])(this.context, nextContext);
			}
		}, {
			key: "getWrappedPureComponent",
			value: function getWrappedPureComponent() {
				return this.refs.pureSeries;
			}
		}, {
			key: "render",
			value: function render() {
				// console.log(PureSeries.defaultProps);
				return _react2["default"].createElement(PureSeries, _extends({ ref: "pureSeries"
				}, this.context, this.props));
			}
		}]);

		return PureCanvasSeries;
	})(_react2["default"].Component);

	;

	PureCanvasSeries.displayName = "pure(" + getDisplayName(PureSeries) + ")";

	PureCanvasSeries.contextTypes = contextShape;
	PureCanvasSeries.yAccessor = PureSeries.yAccessor;

	var defaultProps = {};

	if (PureSeries.defaultProps) {
		Object.keys(PureSeries.defaultProps).filter(function (key) {
			return key === "namespace";
		}).forEach(function (key) {
			defaultProps[key] = PureSeries.defaultProps[key];
		});
		PureCanvasSeries.defaultProps = defaultProps;
	}

	/* Object.keys(PureSeries)
 	.filter((key) => key !== "propTypes")
 	.filter(key => key !== "defaultProps")
 	.filter(key => key !== "displayName")
 	.filter(key => key !== "contextTypes")
 	.filter(key => key !== "childContextTypes")
 	.forEach(key => PureCanvasSeries[key] = PureSeries[key]);*/

	// console.log(Object.keys(PureCanvasSeries))
	return PureCanvasSeries;
}

exports["default"] = pure;
module.exports = exports["default"];
