// react-stockcharts/lib/series/BollingerSeries.js
"use strict";

Object.defineProperty(exports, "__esModule", {
	value: true
});

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

var _get = function get(_x, _x2, _x3) { var _again = true; _function: while (_again) { var object = _x, property = _x2, receiver = _x3; desc = parent = getter = undefined; _again = false; if (object === null) object = Function.prototype; var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { _x = parent; _x2 = property; _x3 = receiver; _again = true; continue _function; } } else if ("value" in desc) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } } };

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var _react = require("react");

var _react2 = _interopRequireDefault(_react);

var _Line = require("./Line");

var _Line2 = _interopRequireDefault(_Line);

var _Area = require("./Area");

var _Area2 = _interopRequireDefault(_Area);

var _wrap = require("./wrap");

var _wrap2 = _interopRequireDefault(_wrap);

var BollingerSeries = (function (_React$Component) {
	_inherits(BollingerSeries, _React$Component);

	function BollingerSeries(props) {
		_classCallCheck(this, BollingerSeries);

		_get(Object.getPrototypeOf(BollingerSeries.prototype), "constructor", this).call(this, props);
		this.yAccessorForTop = this.yAccessorForTop.bind(this);
		this.yAccessorForMiddle = this.yAccessorForMiddle.bind(this);
		this.yAccessorForBottom = this.yAccessorForBottom.bind(this);
		this.yAccessorForScalledBottom = this.yAccessorForScalledBottom.bind(this);
	}

	_createClass(BollingerSeries, [{
		key: "yAccessorForTop",
		value: function yAccessorForTop(d) {
			var yAccessor = this.props.yAccessor;

			return yAccessor(d) && yAccessor(d).top;
		}
	}, {
		key: "yAccessorForMiddle",
		value: function yAccessorForMiddle(d) {
			var yAccessor = this.props.yAccessor;

			return yAccessor(d) && yAccessor(d).middle;
		}
	}, {
		key: "yAccessorForBottom",
		value: function yAccessorForBottom(d) {
			var yAccessor = this.props.yAccessor;

			return yAccessor(d) && yAccessor(d).bottom;
		}
	}, {
		key: "yAccessorForScalledBottom",
		value: function yAccessorForScalledBottom(scale, d) {
			var yAccessor = this.props.yAccessor;

			return scale(yAccessor(d) && yAccessor(d).bottom);
		}
	}, {
		key: "render",
		value: function render() {
			var _props = this.props;
			var xScale = _props.xScale;
			var yScale = _props.yScale;
			var xAccessor = _props.xAccessor;
			var plotData = _props.plotData;
			var type = _props.type;
			var _props2 = this.props;
			var stroke = _props2.stroke;
			var areaClassName = _props2.areaClassName;
			var className = _props2.className;
			var fill = _props2.fill;
			var opacity = _props2.opacity;

			return _react2["default"].createElement(
				"g",
				{ className: className },
				_react2["default"].createElement(_Line2["default"], {
					xScale: xScale, yScale: yScale,
					xAccessor: xAccessor, yAccessor: this.yAccessorForTop,
					plotData: plotData,
					stroke: stroke.top, fill: "none",
					type: type }),
				_react2["default"].createElement(_Line2["default"], {
					xScale: xScale, yScale: yScale,
					xAccessor: xAccessor, yAccessor: this.yAccessorForMiddle,
					plotData: plotData,
					stroke: stroke.middle, fill: "none",
					type: type }),
				_react2["default"].createElement(_Line2["default"], {
					xScale: xScale, yScale: yScale,
					xAccessor: xAccessor, yAccessor: this.yAccessorForBottom,
					plotData: plotData,
					stroke: stroke.bottom, fill: "none",
					type: type }),
				_react2["default"].createElement(_Area2["default"], {
					className: areaClassName,
					xScale: xScale, yScale: yScale,
					xAccessor: xAccessor, yAccessor: this.yAccessorForTop,
					base: this.yAccessorForScalledBottom,
					plotData: plotData,
					stroke: "none", fill: fill, opacity: opacity,
					type: type })
			);
		}
	}]);

	return BollingerSeries;
})(_react2["default"].Component);

BollingerSeries.propTypes = {
	xAccessor: _react2["default"].PropTypes.func,
	yAccessor: _react2["default"].PropTypes.func,
	xScale: _react2["default"].PropTypes.func,
	yScale: _react2["default"].PropTypes.func,
	plotData: _react2["default"].PropTypes.array,
	stroke: _react2["default"].PropTypes.object,
	className: _react2["default"].PropTypes.string,
	areaClassName: _react2["default"].PropTypes.string,
	fill: _react2["default"].PropTypes.string,
	opacity: _react2["default"].PropTypes.number,
	type: _react2["default"].PropTypes.string
};

BollingerSeries.defaultProps = {
	className: "react-stockcharts-bollinger-band-series",
	areaClassName: "react-stockcharts-bollinger-band-series-area",
	stroke: {
		top: "brown",
		middle: "black",
		bottom: "brown"
	},
	fill: "#4682B4",
	opacity: 0.2
};

exports["default"] = (0, _wrap2["default"])(BollingerSeries);
module.exports = exports["default"];
