// react-stockcharts/lib/series/CompareSeries.js
"use strict";

Object.defineProperty(exports, "__esModule", {
	value: true
});

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

var _get = function get(_x, _x2, _x3) { var _again = true; _function: while (_again) { var object = _x, property = _x2, receiver = _x3; desc = parent = getter = undefined; _again = false; if (object === null) object = Function.prototype; var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { _x = parent; _x2 = property; _x3 = receiver; _again = true; continue _function; } } else if ("value" in desc) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } } };

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var _react = require("react");

var _react2 = _interopRequireDefault(_react);

var _Line = require("./Line");

var _Line2 = _interopRequireDefault(_Line);

var _wrap = require("./wrap");

var _wrap2 = _interopRequireDefault(_wrap);

var CompareSeries = (function (_React$Component) {
	_inherits(CompareSeries, _React$Component);

	function CompareSeries() {
		_classCallCheck(this, CompareSeries);

		_get(Object.getPrototypeOf(CompareSeries.prototype), "constructor", this).apply(this, arguments);
	}

	_createClass(CompareSeries, [{
		key: "render",
		value: function render() {
			var props = this.props;
			var className = props.className;
			var compareSeries = props.compareSeries;
			var xScale = props.xScale;
			var yScale = props.yScale;
			var xAccessor = props.xAccessor;
			var plotData = props.plotData;
			var type = props.type;
			var id = props.id;

			var thisSeries = compareSeries.filter(function (each) {
				return each.id === id;
			})[0];
			return _react2["default"].createElement(_Line2["default"], {
				className: className,
				xScale: xScale, yScale: yScale,
				xAccessor: xAccessor, yAccessor: thisSeries.percentYAccessor,
				plotData: plotData,
				stroke: thisSeries.stroke, fill: "none",
				type: type });
		}
	}]);

	return CompareSeries;
})(_react2["default"].Component);

CompareSeries.propTypes = {
	className: _react2["default"].PropTypes.string,
	stroke: _react2["default"].PropTypes.string,
	displayLabel: _react2["default"].PropTypes.string.isRequired,
	id: _react2["default"].PropTypes.number.isRequired
};

CompareSeries.defaultProps = {
	className: "line ",
	namespace: "ReStock.CompareSeries"
};

exports["default"] = (0, _wrap2["default"])(CompareSeries);
module.exports = exports["default"];
