// react-stockcharts/lib/series/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
	value: true
});

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

var _AreaSeries = require("./AreaSeries");

var _AreaSeries2 = _interopRequireDefault(_AreaSeries);

var _LineSeries = require("./LineSeries");

var _LineSeries2 = _interopRequireDefault(_LineSeries);

var _CompareSeries = require("./CompareSeries");

var _CompareSeries2 = _interopRequireDefault(_CompareSeries);

var _CandlestickSeries = require("./CandlestickSeries");

var _CandlestickSeries2 = _interopRequireDefault(_CandlestickSeries);

var _HistogramSeries = require("./HistogramSeries");

var _HistogramSeries2 = _interopRequireDefault(_HistogramSeries);

var _KagiSeries = require("./KagiSeries");

var _KagiSeries2 = _interopRequireDefault(_KagiSeries);

var _PointAndFigureSeries = require("./PointAndFigureSeries");

var _PointAndFigureSeries2 = _interopRequireDefault(_PointAndFigureSeries);

var _RenkoSeries = require("./RenkoSeries");

var _RenkoSeries2 = _interopRequireDefault(_RenkoSeries);

var _MACDSeries = require("./MACDSeries");

var _MACDSeries2 = _interopRequireDefault(_MACDSeries);

var _BollingerSeries = require("./BollingerSeries");

var _BollingerSeries2 = _interopRequireDefault(_BollingerSeries);

var _RSISeries = require("./RSISeries");

var _RSISeries2 = _interopRequireDefault(_RSISeries);

var _StochasticSeries = require("./StochasticSeries");

var _StochasticSeries2 = _interopRequireDefault(_StochasticSeries);

var _Area = require("./Area");

var _Area2 = _interopRequireDefault(_Area);

var _Line = require("./Line");

var _Line2 = _interopRequireDefault(_Line);

var _StraightLine = require("./StraightLine");

var _StraightLine2 = _interopRequireDefault(_StraightLine);

var _wrap = require("./wrap");

var _wrap2 = _interopRequireDefault(_wrap);

exports["default"] = {
	AreaSeries: _AreaSeries2["default"],
	LineSeries: _LineSeries2["default"],
	CompareSeries: _CompareSeries2["default"],
	CandlestickSeries: _CandlestickSeries2["default"],
	HistogramSeries: _HistogramSeries2["default"],
	KagiSeries: _KagiSeries2["default"],
	PointAndFigureSeries: _PointAndFigureSeries2["default"],
	RenkoSeries: _RenkoSeries2["default"],
	MACDSeries: _MACDSeries2["default"],
	BollingerSeries: _BollingerSeries2["default"],
	RSISeries: _RSISeries2["default"],
	StochasticSeries: _StochasticSeries2["default"],
	Area: _Area2["default"],
	Line: _Line2["default"],
	StraightLine: _StraightLine2["default"],
	wrap: _wrap2["default"]
};
module.exports = exports["default"];
