// react-stockcharts/lib/series/RSISeries.js
"use strict";

Object.defineProperty(exports, "__esModule", {
	value: true
});

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

var _get = function get(_x, _x2, _x3) { var _again = true; _function: while (_again) { var object = _x, property = _x2, receiver = _x3; desc = parent = getter = undefined; _again = false; if (object === null) object = Function.prototype; var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { _x = parent; _x2 = property; _x3 = receiver; _again = true; continue _function; } } else if ("value" in desc) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } } };

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var _react = require("react");

var _react2 = _interopRequireDefault(_react);

var _Line = require("./Line");

var _Line2 = _interopRequireDefault(_Line);

var _StraightLine = require("./StraightLine");

var _StraightLine2 = _interopRequireDefault(_StraightLine);

var _wrap = require("./wrap");

var _wrap2 = _interopRequireDefault(_wrap);

var RSISeries = (function (_React$Component) {
	_inherits(RSISeries, _React$Component);

	function RSISeries() {
		_classCallCheck(this, RSISeries);

		_get(Object.getPrototypeOf(RSISeries.prototype), "constructor", this).apply(this, arguments);
	}

	_createClass(RSISeries, [{
		key: "render",
		value: function render() {
			var _props = this.props;
			var className = _props.className;
			var indicator = _props.indicator;
			var xScale = _props.xScale;
			var yScale = _props.yScale;
			var xAccessor = _props.xAccessor;
			var yAccessor = _props.yAccessor;
			var plotData = _props.plotData;
			var stroke = _props.stroke;
			var type = _props.type;

			var options = indicator.options();
			return _react2["default"].createElement(
				"g",
				{ className: className },
				_react2["default"].createElement(_Line2["default"], {
					className: className,
					xScale: xScale, yScale: yScale,
					xAccessor: xAccessor, yAccessor: yAccessor,
					plotData: plotData,
					stroke: stroke.line, fill: "none",
					type: type }),
				RSISeries.getHorizontalLine(this.props, options.overSold, stroke.top),
				RSISeries.getHorizontalLine(this.props, 50, stroke.middle),
				RSISeries.getHorizontalLine(this.props, options.overBought, stroke.bottom)
			);
		}
	}]);

	return RSISeries;
})(_react2["default"].Component);

RSISeries.getHorizontalLine = function (props, yValue, stroke) {
	/* eslint-disable react/prop-types */
	var xScale = props.xScale;
	var yScale = props.yScale;
	var xAccessor = props.xAccessor;
	var yAccessor = props.yAccessor;
	var plotData = props.plotData;
	var type = props.type;

	/* eslint-enable react/prop-types */

	return _react2["default"].createElement(_StraightLine2["default"], {
		stroke: stroke, opacity: 0.3, type: type,
		xScale: xScale, yScale: yScale,
		xAccessor: xAccessor, yAccessor: yAccessor,
		plotData: plotData,
		yValue: yValue });
};

RSISeries.propTypes = {
	className: _react2["default"].PropTypes.string,

	indicator: _react2["default"].PropTypes.func,
	xScale: _react2["default"].PropTypes.func,
	yScale: _react2["default"].PropTypes.func,
	xAccessor: _react2["default"].PropTypes.func,
	yAccessor: _react2["default"].PropTypes.func,
	plotData: _react2["default"].PropTypes.array,
	stroke: _react2["default"].PropTypes.object,
	type: _react2["default"].PropTypes.string
};

RSISeries.defaultProps = {
	className: "react-stockcharts-rsi-series",
	stroke: {
		line: "#000000",
		top: "brown",
		middle: "black",
		bottom: "brown"
	}
};

exports["default"] = (0, _wrap2["default"])(RSISeries);
module.exports = exports["default"];
