// react-stockcharts/lib/transforms/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
	value: true
});

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj["default"] = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

var _DummyTransformer = require("./DummyTransformer");

var _DummyTransformer2 = _interopRequireDefault(_DummyTransformer);

var _StockscaleTransformer = require("./StockscaleTransformer");

var _StockscaleTransformer2 = _interopRequireDefault(_StockscaleTransformer);

var _HeikinAshiTransformer = require("./HeikinAshiTransformer");

var _HeikinAshiTransformer2 = _interopRequireDefault(_HeikinAshiTransformer);

var _KagiTransformer = require("./KagiTransformer");

var _KagiTransformer2 = _interopRequireDefault(_KagiTransformer);

var _RenkoTransformer = require("./RenkoTransformer");

var _RenkoTransformer2 = _interopRequireDefault(_RenkoTransformer);

var _PointAndFigureTransformer = require("./PointAndFigureTransformer");

var _PointAndFigureTransformer2 = _interopRequireDefault(_PointAndFigureTransformer);

var _defaultOptions = require("./defaultOptions");

var defaultOptions = _interopRequireWildcard(_defaultOptions);

exports["default"] = {
	DummyTransformer: _DummyTransformer2["default"],
	StockscaleTransformer: _StockscaleTransformer2["default"],
	HeikinAshiTransformer: _HeikinAshiTransformer2["default"],
	KagiTransformer: _KagiTransformer2["default"],
	RenkoTransformer: _RenkoTransformer2["default"],
	PointAndFigureTransformer: _PointAndFigureTransformer2["default"],
	defaultOptions: defaultOptions
};
module.exports = exports["default"];
