// react-stockcharts/lib/transforms/RenkoTransformer.js
"use strict";

Object.defineProperty(exports, "__esModule", {
	value: true
});

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

var _utilsATRCalculator = require("../utils/ATRCalculator");

var _utilsATRCalculator2 = _interopRequireDefault(_utilsATRCalculator);

var _objectAssign = require("object-assign");

var _objectAssign2 = _interopRequireDefault(_objectAssign);

var _defaultOptions = require("./defaultOptions");

function RenkoTransformer() {
	var newOptions;
	function transform(rawData /* , interval */) {
		var _newOptions = newOptions;
		var reversalType = _newOptions.reversalType;
		var fixedBrickSize = _newOptions.fixedBrickSize;
		var dateAccessor = _newOptions.dateAccessor;
		var dateMutator = _newOptions.dateMutator;
		var indexAccessor = _newOptions.indexAccessor;
		var indexMutator = _newOptions.indexMutator;
		var period = _newOptions.period;
		var source = _newOptions.source;

		var brickSize, pricingMethod;
		if (reversalType === "ATR") {
			(0, _utilsATRCalculator2["default"])(rawData.D, period);
			brickSize = function (d) {
				return d["atr" + period];
			};
		} else {
			brickSize = function () {
				return fixedBrickSize;
			};
		}
		if (source === "hi/lo") {
			pricingMethod = function (d) {
				return { high: d.high, low: d.low };
			};
		} else {
			pricingMethod = function (d) {
				return { high: d.close, low: d.close };
			};
		}

		var renkoData = [];

		var index = 0,
		    prevBrickClose = rawData.D[index].open,
		    prevBrickOpen = rawData.D[index].open;
		var brick = {},
		    direction = 0;

		rawData.D.forEach(function (d) {
			if (brick.from === undefined) {
				brick.high = d.high;
				brick.low = d.low;
				brick.startOfYear = d.startOfYear;
				brick.startOfQuarter = d.startOfQuarter;
				brick.startOfMonth = d.startOfMonth;
				brick.startOfWeek = d.startOfWeek;

				brick.from = indexAccessor(d);
				brick.fromDate = dateAccessor(d);
				indexMutator(brick, index++);
				dateMutator(brick, dateAccessor(d));
			}
			brick.volume = (brick.volume || 0) + d.volume;

			var prevCloseToHigh = prevBrickClose - pricingMethod(d).high,
			    prevCloseToLow = prevBrickClose - pricingMethod(d).low,
			    prevOpenToHigh = prevBrickOpen - pricingMethod(d).high,
			    prevOpenToLow = prevBrickOpen - pricingMethod(d).low,
			    priceMovement = Math.min(Math.abs(prevCloseToHigh), Math.abs(prevCloseToLow), Math.abs(prevOpenToHigh), Math.abs(prevOpenToLow));

			brick.high = Math.max(brick.high, d.high);
			brick.low = Math.min(brick.low, d.low);

			if (!brick.startOfYear) {
				brick.startOfYear = d.startOfYear;
				if (brick.startOfYear) {
					dateMutator(brick, dateAccessor(d));
					// brick.displayDate = d.displayDate;
				}
			}

			if (!brick.startOfQuarter) {
				brick.startOfQuarter = d.startOfQuarter;
				if (brick.startOfQuarter && !brick.startOfYear) {
					dateMutator(brick, dateAccessor(d));
					// brick.displayDate = d.displayDate;
				}
			}

			if (!brick.startOfMonth) {
				brick.startOfMonth = d.startOfMonth;
				if (brick.startOfMonth && !brick.startOfQuarter) {
					dateMutator(brick, dateAccessor(d));
					// brick.displayDate = d.displayDate;
				}
			}
			if (!brick.startOfWeek) {
				brick.startOfWeek = d.startOfWeek;
				if (brick.startOfWeek && !brick.startOfMonth) {
					dateMutator(brick, dateAccessor(d));
					// brick.displayDate = d.displayDate;
				}
			}

			// d.brick = JSON.stringify(brick);
			if (brickSize(d)) {
				var noOfBricks = Math.floor(priceMovement / brickSize(d));

				brick.open = Math.abs(prevCloseToHigh) < Math.abs(prevOpenToHigh) || Math.abs(prevCloseToLow) < Math.abs(prevOpenToLow) ? prevBrickClose : prevBrickOpen;

				if (noOfBricks >= 1) {
					var j = 0;
					for (j = 0; j < noOfBricks; j++) {
						brick.close = brick.open < pricingMethod(d).high ?
						// if brick open is less than current price it means it is green/hollow brick
						brick.open + brickSize(d) : brick.open - brickSize(d);
						direction = brick.close > brick.open ? 1 : -1;
						brick.direction = direction;
						brick.to = indexAccessor(d);
						brick.toDate = dateAccessor(d);
						// brick.diff = brick.open - brick.close;
						// brick.atr = d.atr;
						brick.fullyFormed = true;
						renkoData.push(brick);

						prevBrickClose = brick.close;
						prevBrickOpen = brick.open;

						var newBrick = {
							high: brick.high,
							low: brick.low,
							open: brick.close,
							startOfYear: false,
							startOfMonth: false,
							startOfQuarter: false,
							startOfWeek: false
						};
						brick = newBrick;
						brick.from = indexAccessor(d);
						brick.fromDate = dateAccessor(d);
						indexMutator(brick, index + j);
						dateMutator(brick, dateAccessor(d));
						brick.volume = (brick.volume || 0) + d.volume;
					}
					index = index + j - 1;
					brick = {};
				} else {
					if (indexAccessor(d) === rawData.D.length - 1) {
						brick.close = direction > 0 ? pricingMethod(d).high : pricingMethod(d).low;
						brick.to = indexAccessor(d);
						brick.toDate = dateAccessor(d);
						dateMutator(brick, dateAccessor(d));

						brick.fullyFormed = false;
						renkoData.push(brick);
					}
				}
			}
		});
		return { "D": renkoData };
	};

	transform.options = function (opt) {
		newOptions = (0, _objectAssign2["default"])({}, _defaultOptions.Renko, opt);
		return newOptions;
	};
	return transform;
}

exports["default"] = RenkoTransformer;
module.exports = exports["default"];
