// react-stockcharts/lib/interactive/ClickCallback.js
"use strict";

Object.defineProperty(exports, "__esModule", {
	value: true
});

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

var _get = function get(_x, _x2, _x3) { var _again = true; _function: while (_again) { var object = _x, property = _x2, receiver = _x3; desc = parent = getter = undefined; _again = false; if (object === null) object = Function.prototype; var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { _x = parent; _x2 = property; _x3 = receiver; _again = true; continue _function; } } else if ("value" in desc) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } } };

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var _react = require("react");

var _react2 = _interopRequireDefault(_react);

var _makeInteractive = require("./makeInteractive");

var _makeInteractive2 = _interopRequireDefault(_makeInteractive);

var ClickCallback = (function (_React$Component) {
	_inherits(ClickCallback, _React$Component);

	function ClickCallback(props) {
		_classCallCheck(this, ClickCallback);

		_get(Object.getPrototypeOf(ClickCallback.prototype), "constructor", this).call(this, props);
		this.onClick = this.onClick.bind(this);
	}

	_createClass(ClickCallback, [{
		key: "onClick",
		value: function onClick(chartId, xAccessor, interactive, _ref /* , e */) {
			var mouseXY = _ref.mouseXY;
			var currentItem = _ref.currentItem;
			var chartData = _ref.chartData;
			var onClick = this.props.onClick;
			var yScale = chartData.plot.scales.yScale;

			var yValue = yScale.invert(mouseXY[1]);
			var xValue = xAccessor(currentItem);
			onClick({
				xy: [xValue, yValue],
				mouseXY: mouseXY,
				currentItem: currentItem
			});
			return interactive; // returning the same input to indicate that the state of the chart is not changing
		}
	}, {
		key: "render",
		value: function render() {
			return null;
		}
	}]);

	return ClickCallback;
})(_react2["default"].Component);

ClickCallback.drawOnCanvas = function () {};

ClickCallback.propTypes = {
	onClick: _react2["default"].PropTypes.func.isRequired
};

ClickCallback.defaultProps = {
	onClick: function onClick(e) {
		console.log(e);
	}
};

exports["default"] = (0, _makeInteractive2["default"])(ClickCallback, ["click"], {});
module.exports = exports["default"];
