// react-stockcharts/index.js
"use strict";

// common components
Object.defineProperty(exports, "__esModule", {
	value: true
});

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj["default"] = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

var _libChartCanvas = require("./lib/ChartCanvas");

var _libChartCanvas2 = _interopRequireDefault(_libChartCanvas);

var _libChart = require("./lib/Chart");

var _libChart2 = _interopRequireDefault(_libChart);

var _libDataSeries = require("./lib/DataSeries");

var _libDataSeries2 = _interopRequireDefault(_libDataSeries);

var _libBackgroundText = require("./lib/BackgroundText");

var _libBackgroundText2 = _interopRequireDefault(_libBackgroundText);

// interaction components

var _libEventCapture = require("./lib/EventCapture");

var _libEventCapture2 = _interopRequireDefault(_libEventCapture);

// chart types & Series

var _libSeries = require("./lib/series");

var _libSeries2 = _interopRequireDefault(_libSeries);

var _libCoordinates = require("./lib/coordinates");

var _libCoordinates2 = _interopRequireDefault(_libCoordinates);

var _libIndicator = require("./lib/indicator");

var _libIndicator2 = _interopRequireDefault(_libIndicator);

var _libTransforms = require("./lib/transforms");

var _libTransforms2 = _interopRequireDefault(_libTransforms);

var _libAxes = require("./lib/axes");

var _libAxes2 = _interopRequireDefault(_libAxes);

var _libTooltip = require("./lib/tooltip");

var _libTooltip2 = _interopRequireDefault(_libTooltip);

var _libHelper = require("./lib/helper");

var _libHelper2 = _interopRequireDefault(_libHelper);

var _libInteractive = require("./lib/interactive");

var _libInteractive2 = _interopRequireDefault(_libInteractive);

var _libUtilsUtils = require("./lib/utils/utils");

var Utils = _interopRequireWildcard(_libUtilsUtils);

var version = "0.3.1";

exports["default"] = {
	ChartCanvas: _libChartCanvas2["default"],
	Chart: _libChart2["default"],
	DataSeries: _libDataSeries2["default"],
	EventCapture: _libEventCapture2["default"],
	BackgroundText: _libBackgroundText2["default"],
	series: _libSeries2["default"],
	coordinates: _libCoordinates2["default"],
	indicator: _libIndicator2["default"],
	transforms: _libTransforms2["default"],
	axes: _libAxes2["default"],
	tooltip: _libTooltip2["default"],
	helper: _libHelper2["default"],
	interactive: _libInteractive2["default"],
	version: version,
	Utils: Utils
};
module.exports = exports["default"];
