// react-stockcharts/lib/coordinates/EdgeContainer.js
"use strict";

Object.defineProperty(exports, "__esModule", {
	value: true
});

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

var _get = function get(_x, _x2, _x3) { var _again = true; _function: while (_again) { var object = _x, property = _x2, receiver = _x3; desc = parent = getter = undefined; _again = false; if (object === null) object = Function.prototype; var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { _x = parent; _x2 = property; _x3 = receiver; _again = true; continue _function; } } else if ("value" in desc) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } } };

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var _react = require("react");

var _react2 = _interopRequireDefault(_react);

var _utilsUtils = require("../utils/utils");

var _objectAssign = require("object-assign");

var _objectAssign2 = _interopRequireDefault(_objectAssign);

var _utilsPureComponent = require("../utils/PureComponent");

var _utilsPureComponent2 = _interopRequireDefault(_utilsPureComponent);

var EdgeContainer = (function (_PureComponent) {
	_inherits(EdgeContainer, _PureComponent);

	function EdgeContainer() {
		_classCallCheck(this, EdgeContainer);

		_get(Object.getPrototypeOf(EdgeContainer.prototype), "constructor", this).apply(this, arguments);
	}

	_createClass(EdgeContainer, [{
		key: "render",
		value: function render() {
			var _this = this;

			var children = _react2["default"].Children.map(this.props.children, function (child) {
				var newChild = (0, _utilsUtils.isReactVersion13)() ? _react2["default"].withContext(_this.context, function () {
					return _react2["default"].createElement(child.type, (0, _objectAssign2["default"])({ key: child.key, ref: child.ref }, child.props));
				}) : child;
				// React.createElement(child.type, objectAssign({ key: child.key, ref: child.ref}, child.props));
				return newChild;
			});
			return _react2["default"].createElement(
				"g",
				null,
				children
			);
		}
	}]);

	return EdgeContainer;
})(_utilsPureComponent2["default"]);

EdgeContainer.contextTypes = {
	width: _react2["default"].PropTypes.number.isRequired,
	chartData: _react2["default"].PropTypes.array.isRequired,
	// currentItems: React.PropTypes.array.isRequired,
	axesCanvasContext: _react2["default"].PropTypes.object,
	type: _react2["default"].PropTypes.string,
	margin: _react2["default"].PropTypes.object.isRequired
};

// secretToSuperFastCanvasDraw: React.PropTypes.array.isRequired,
exports["default"] = EdgeContainer;
module.exports = exports["default"];
