// react-stockcharts/lib/coordinates/VerticalMousePointer.js
"use strict";

Object.defineProperty(exports, "__esModule", {
	value: true
});

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

var _get = function get(_x, _x2, _x3) { var _again = true; _function: while (_again) { var object = _x, property = _x2, receiver = _x3; desc = parent = getter = undefined; _again = false; if (object === null) object = Function.prototype; var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { _x = parent; _x2 = property; _x3 = receiver; _again = true; continue _function; } } else if ("value" in desc) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } } };

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var _react = require("react");

var _react2 = _interopRequireDefault(_react);

var _EdgeCoordinate = require("./EdgeCoordinate");

var _EdgeCoordinate2 = _interopRequireDefault(_EdgeCoordinate);

var VerticalMousePointer = (function (_React$Component) {
	_inherits(VerticalMousePointer, _React$Component);

	function VerticalMousePointer() {
		_classCallCheck(this, VerticalMousePointer);

		_get(Object.getPrototypeOf(VerticalMousePointer.prototype), "constructor", this).apply(this, arguments);
	}

	_createClass(VerticalMousePointer, [{
		key: "shouldComponentUpdate",
		value: function shouldComponentUpdate(nextProps) {
			return nextProps.mouseXY !== this.props.mouseXY;
		}
	}, {
		key: "render",
		value: function render() {
			return _react2["default"].createElement(
				"g",
				{ className: "crosshair " },
				_react2["default"].createElement(_EdgeCoordinate2["default"], {
					type: "vertical",
					className: "horizontal",
					show: true,
					x1: this.props.mouseXY[0], y1: 0,
					x2: this.props.mouseXY[0], y2: this.props.height,
					coordinate: this.props.xDisplayValue,
					edgeAt: this.props.height,
					orient: "bottom" })
			);
		}
	}]);

	return VerticalMousePointer;
})(_react2["default"].Component);

VerticalMousePointer.propTypes = {
	height: _react2["default"].PropTypes.number.isRequired,
	mouseXY: _react2["default"].PropTypes.array.isRequired,
	xDisplayValue: _react2["default"].PropTypes.string.isRequired
};

VerticalMousePointer.defaultProps = { namespace: "ReStock.VerticalMousePointer" };

VerticalMousePointer.yAccessor = function (d) {
	return { open: d.open, high: d.high, low: d.low, close: d.close };
};

exports["default"] = VerticalMousePointer;
module.exports = exports["default"];
