// react-stockcharts/lib/indicator/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
	value: true
});

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj["default"] = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

var _MACDIndicator = require("./MACDIndicator");

var _MACDIndicator2 = _interopRequireDefault(_MACDIndicator);

var _EMAIndicator = require("./EMAIndicator");

var _EMAIndicator2 = _interopRequireDefault(_EMAIndicator);

var _SMAIndicator = require("./SMAIndicator");

var _SMAIndicator2 = _interopRequireDefault(_SMAIndicator);

var _BollingerBandIndicator = require("./BollingerBandIndicator");

var _BollingerBandIndicator2 = _interopRequireDefault(_BollingerBandIndicator);

var _RSIIndicator = require("./RSIIndicator");

var _RSIIndicator2 = _interopRequireDefault(_RSIIndicator);

var _FullStochasticOscillator = require("./FullStochasticOscillator");

var _FullStochasticOscillator2 = _interopRequireDefault(_FullStochasticOscillator);

var _defaultOptions = require("./defaultOptions");

var defaultOptions = _interopRequireWildcard(_defaultOptions);

exports["default"] = {
	MACD: _MACDIndicator2["default"],
	EMA: _EMAIndicator2["default"],
	SMA: _SMAIndicator2["default"],
	BollingerBand: _BollingerBandIndicator2["default"],
	RSI: _RSIIndicator2["default"],
	FullStochasticOscillator: _FullStochasticOscillator2["default"],
	defaultOptions: defaultOptions
};
module.exports = exports["default"];
